src/DomainHost.sol

```diff
@@ -0,0 +1,449 @@
+            vat.frob(ilk, address(this), address(this), address(this), int256(dlineWad), int256(dlineWad));
+            daiJoin.exit(escrow, dlineWad);
+
+            grain += dlineWad;
+        }
+
+        line = rad;
+
+        _rid = rid++;
+
+        emit Lift(wad);
+    }
+
+    /// @notice Withdraw pre-mint DAI from the remote domain
+    /// @param _lid Local ordering id
+    /// @param wad The amount of DAI to release [WAD]
+    function _release(uint256 _lid, uint256 wad) internal ordered(_lid) {
+        require(vat.live() == 1, "DomainHost/vat-not-live");
+
+        // Fix any permissionless repays that may have occurred
+        (uint256 ink, uint256 art) = vat.urns(ilk, address(this));
+        if (art < ink) {
+            address _vow = vow;
+            uint256 diff = ink - art;
+            vat.suck(_vow, _vow, diff * RAY); // This needs to be done to make sure we can deduct sin[vow] and vice in the next call
+            vat.grab(ilk, address(this), address(this), _vow, 0, _int256(diff));
+        }
+
+        require(dai.transferFrom(escrow, address(this), wad), "DomainHost/transfer-failed");
+        daiJoin.join(address(this), wad);
+        int256 repay = -_int256(wad);
+        vat.frob(ilk, address(this), address(this), address(this), repay, repay);
+        vat.slip(ilk, address(this), repay);
+
+        grain -= wad;
+
+        emit Release(wad);
+    }
+
+    /// @notice Guest is pushing a surplus
+    /// @param _lid Local ordering id
+    /// @param wad The amount of DAI to send to the vow [WAD]
+    function _surplus(uint256 _lid, uint256 wad) internal ordered(_lid) {
+        dai.transferFrom(address(escrow), address(this), uint256(wad));
+        daiJoin.join(address(vow), uint256(wad));
+
+        emit Surplus(wad);
+    }
+
+    /// @notice Guest is pushing a deficit
+    /// @param _lid Local ordering id
+    /// @param wad The amount of DAI to account as dsin [WAD]
+    function _deficit(uint256 _lid, uint256 wad) internal ordered(_lid) {
+        dsin += wad;
+
+        emit Deficit(wad);
+    }
+
+    /// @notice Move bad debt from the remote domain into the local vow
+    /// @dev This is a potentially dangerous operation as a malicious domain can drain the entire surplus buffer
+    /// Because of this we require an authed party to perform this operation
+    function _rectify() internal auth returns (uint256 _rid, uint256 _wad) {
+        require(vat.live() == 1, "DomainHost/vat-not-live");
+
+        _wad = dsin;
+        require(_wad > 0, "DomainHost/no-sin");
+        vat.suck(vow, address(this), _wad * RAY);
+        daiJoin.exit(address(escrow), _wad);
+        dsin = 0;
+
+        _rid = rid++;
+
+        emit Rectify(_wad);
+    }
+
+    /// @notice Initiate shutdown for this domain
+    /// @dev This will trigger the end module on the remote domain
+    function _cage() internal returns (uint256 _rid) {
```

Comment:
I think `cage`, `rectify` and `lift` will also need `undo` methods.

Reply:
I thought about this, and I don't think this is necessary. In the scenario where a particular domain begins censoring Maker we will probably just want to shut the whole thing down if we can.

In the event we think we can recover (maybe a one-off censor?) I would like to fix the state with manual spell. This hasn't been added yet, but the ability to arbitrarily modify the state in dss-bridge makes sense to recover from bad state.

Reply:
A manual state modification function would make sense, but not sure it can be implemented in a way that guarantees that user deposits in the escrow can't be seized. 

Also, note that the L2 part of a cross-chain tx failing may be rather common and won't necessarily require sequencer censorship. For example on zkSync, the fee paid to the operator will depend on the L1 gas price. If the L1 gas price surges shortly after a L1>L2 message is submitted, the fee paid may no longer be sufficient to cover the cost of the L2 tx which may revert as a result.

Reply:
Hmm ok, didn't realize you can have a tx fail for non-censorship reasons.

Reply:
Funds in the escrow are already at mercy of Governance, as pause proxy can actually give allowance to it to new contracts (or to itself).

Reply:
I think we just want to allow `file()` on all these parameters in case something unexpected happens. As Gonza said the escrow is already completely at the mercy of governance so we should rely on ESM for governance attack as per usual.

Reply:
Added in PR: https://github.com/makerdao/dss-bridge/pull/24